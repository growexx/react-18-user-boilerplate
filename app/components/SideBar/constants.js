/*
 * Sidebar Constants
 * This file contains constants used in Sidebar component.
 */

import React from 'react';
import {
  UserOutlined,
  SmileOutlined,
  LoadingOutlined,
  ExportOutlined,
  NumberOutlined,
  FormOutlined,
  LockOutlined,
  PieChartOutlined,
  MessageOutlined,
  ShoppingOutlined,
  SearchOutlined,
  TableOutlined,
} from '@ant-design/icons';
import {
  ROUTES,
  ROLE_BASED_SIDEBAR_MENU,
  ROLES,
  RESTRICTED_ROUTES,
} from 'containers/constants';

export const MenuItems = [
  {
    to: ROUTES.HOME,
    tabName: 'Home',
    icon: <SmileOutlined />,
  },
  {
    to: ROUTES.LOADER,
    tabName: 'Loader Demo',
    icon: <LoadingOutlined />,
  },
  {
    to: ROUTES.EXPORT_DATA,
    tabName: 'Export Data To CSV',
    icon: <ExportOutlined />,
  },
  {
    to: ROUTES.NUMERAL_CONVERTER,
    tabName: 'Number Conversion Demo',
    icon: <NumberOutlined />,
  },
  {
    to: ROUTES.USERS,
    tabName: 'User Management',
    icon: <UserOutlined />,
  },
  {
    to: ROUTES.GITHUB_SEARCH,
    tabName: 'Github Search',
    icon: <SearchOutlined />,
  },
  {
    to: ROUTES.TEST_ADMIN_PAGE,
    tabName: 'Admin Page',
    icon: <LockOutlined />,
  },
  {
    to: ROUTES.CHARTS,
    tabName: 'Charts',
    icon: <PieChartOutlined />,
  },
  {
    to: ROUTES.MULTI_TAB_SUPPORT,
    tabName: 'Multi Tab Communication',
    icon: <MessageOutlined />,
  },
  {
    to: ROUTES.PRODUCTS,
    tabName: 'Products',
    icon: <ShoppingOutlined />,
  },
  {
    to: ROUTES.REACT_HOOK_FORM,
    tabName: 'React Hook Form',
    icon: <FormOutlined />,
  },
  {
    to: ROUTES.AG_GRID,
    tabName: 'AG Grid',
    icon: <TableOutlined />,
  },
];

/**
 * Filters Sidebar menu based on role
 * @param {string} role
 */
export const GET_FILTERED_MENU_ITEM = role =>
  MenuItems.filter(item =>
    // Check if route is private then role has access to it
    RESTRICTED_ROUTES.includes(item.to)
      ? ROLE_BASED_SIDEBAR_MENU[role || ROLES.USER].includes(item.to)
      : true,
  );

// component name as ga label
export const GA_LABEL_SIDEBAR = 'SideBar';
