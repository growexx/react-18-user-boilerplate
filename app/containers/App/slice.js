import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  appLoading: false,
};

export const appSlice = createSlice({
  name: 'app',
  initialState,
  reducers: {
    loadApp: (state, action) => ({
      ...state,
      appLoading: action.payload,
    }),
  },
});

export const { loadApp } = appSlice.actions;

export default appSlice.reducer;
