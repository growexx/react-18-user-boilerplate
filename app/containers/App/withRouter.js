/* eslint-disable react/function-component-definition */
import React from 'react';
import { useNavigate } from 'react-router-dom';

export const withRouter = Component => props => {
  const navigate = useNavigate();

  return <Component navigate={navigate} {...props} />;
};
