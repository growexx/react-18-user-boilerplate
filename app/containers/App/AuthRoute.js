/* eslint-disable space-before-function-paren */
import React from 'react';
import { Outlet, Navigate } from 'react-router';
import { userExists } from 'utils/Helper';
import { ROUTES } from '../constants';

function AuthRoute() {
  return !userExists() ? <Outlet /> : <Navigate to={ROUTES.HOME} />;
}

export default AuthRoute;
