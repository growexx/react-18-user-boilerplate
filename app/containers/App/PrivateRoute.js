import React from 'react';
import { Outlet, Navigate } from 'react-router';
import { userExists } from 'utils/Helper';
import { ROUTES } from '../constants';

function PrivateRoute() {
  return userExists() ? <Outlet /> : <Navigate to={ROUTES.LOGIN} />;
}

export default PrivateRoute;
