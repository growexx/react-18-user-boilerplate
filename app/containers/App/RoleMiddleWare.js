/* eslint-disable no-unused-vars */
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Outlet } from 'react-router-dom';
/* import { Spin } from 'antd'; */
import {
  RESTRICTED_ROUTES,
  ROLE_BASED_ROUTE_ACCESS,
  ROLE_BASED_DEFAULT_ROUTE,
  ROUTES,
} from '../constants';
import { USER_DATA_KEY } from '../../utils/constants';
import StorageService from '../../utils/StorageService';
import { loginSuccessResponse } from '../Auth/Login/stub/login.stub';
import { withRouter } from './withRouter';

export function getDerivedStateFromProps(props, state) {
  // Only check if route is restricted or not
  // Only Reset isRestrictedRoute if userData is not fetched
  if (!state.userData || !Object.keys(state.userData).length) {
    const isRestrictedRoute = RESTRICTED_ROUTES.includes(props.path);

    return {
      ...state,
      isRestrictedRoute,
      loader: isRestrictedRoute,
    };
  }
  return state;
}

function RoleMiddleWare(props) {
  /**
   * NOTE: for actual implementation of the component, remove comments with following note  "ACTUAL API INTEGRATION CODE" and remove demo code
   */
  const [loader, setLoader] = useState(true);
  const [userData, setUserData] = useState({});
  const [isRestrictedRoute, setIsRestrictedRoute] = useState(true);
  const { component: Component, ...rest } = props;

  // Get role routes
  const doesRoleHaveRouteAccess = (role, route) =>
    (ROLE_BASED_ROUTE_ACCESS[role || 'user'] || []).includes(route);

  /**
   * Fetch user data and store in local state
   */
  /* ACTUAL API INTEGRATION CODE
  const fetchUserRole = () => {
    const data = {
      method: 'GET',
    };
    const requestURL = `${API_ENDPOINTS.USER_DETAILS_API}`;
    request(requestURL, data).then(response => {
      if (response.status) {
        // Save in local storage
        StorageService.set(USER_DATA_KEY, response.data);
        setUserData(response.data);
        takeDecision((response.data && response.data.role) || '');
      } else {
        logout();
        props.navigate('/login');
      }
    });
  };
  */

  /**
   * Takes decision and stop the loader
   */
  const takeDecision = role => {
    setLoader(false);
    /**
     * doesRoleHaveRouteAccess (role, path)
     * Check which path do you need to pass based on paths defined in constant file
     */
    if (!doesRoleHaveRouteAccess(role, props.path)) {
      if (props.showError) {
        props.navigate(ROUTES.UNAUTHORIZED);
      } else {
        props.navigate(ROLE_BASED_DEFAULT_ROUTE[role]);
      }
    }
  };

  // On component load get user data
  useEffect(() => {
    // ACTUAL API INTEGRATION CODE
    // if (isRestrictedRoute) {
    //   fetchUserRole();
    // }

    const response = loginSuccessResponse;
    StorageService.set(USER_DATA_KEY, response.data);
    setUserData(response.data);
    takeDecision((response.data && response.data.role) || '');
  }, []);

  return (
    <>
      <Outlet {...rest} Component={null} userData={userData} />

      {/* ACTUAL API INTEGRATION CODE */}
      {/* {loader ? <Spin spinning={loader} size="large" /> : (
        <Outlet {...rest} Component={null} userData={userData} />
      )} */}
    </>
  );
}

RoleMiddleWare.propTypes = {
  history: PropTypes.any,
  childProps: PropTypes.any,
  component: PropTypes.any,
  path: PropTypes.string,
  navigate: PropTypes.func,
  showError: PropTypes.bool,
};

export default withRouter(RoleMiddleWare);
