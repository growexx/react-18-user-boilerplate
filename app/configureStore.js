import { configureStore } from '@reduxjs/toolkit';
import { createBrowserHistory } from 'history';
import { createReduxHistoryContext } from 'redux-first-history';

import { repoApi } from './containers/HomePage/reposApiSlice';
import { appSlice } from './containers/App/slice';
import { apiSlice } from './apiSlice';
import languageSlice from './containers/LanguageProvider/slice';
import loginSlice from './containers/Auth/Login/slice';
import { authApi } from './containers/Auth/Login/authApiSlice';
import twoFactorAuthSlice from './containers/Auth/TwoFactorAuthentication/slice';

const { createReduxHistory, routerMiddleware, routerReducer } =
  createReduxHistoryContext({ history: createBrowserHistory() });

const store = configureStore({
  reducer: {
    router: routerReducer,
    app: appSlice,
    language: languageSlice,
    login: loginSlice,
    twoFactorAuth: twoFactorAuthSlice,
    repos: repoApi.reducer,
    auth: authApi.reducer,
    [apiSlice.reducerPath]: apiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => [
    ...getDefaultMiddleware(),
    apiSlice.middleware,
    routerMiddleware,
  ],
  devTools: process.env.NODE_ENV !== 'production',
});

const history = createReduxHistory(store);

export { store, history };
