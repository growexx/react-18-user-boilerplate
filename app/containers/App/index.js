/* eslint-disable react/no-array-index-key */
/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import React, { useEffect } from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Routes, Route } from 'react-router-dom';
import HomePage from 'containers/HomePage/Loadable';
import Profile from 'containers/Profile/Loadable';
import FeaturePage from 'containers/FeaturePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import UnauthorizedPage from 'containers/UnauthorizedPage/Loadable';
import Login from 'containers/Auth/Login/Loadable';
import TwoFactorAuthentication from 'containers/Auth/TwoFactorAuthentication/Loadable';
import Logout from 'containers/Auth/Logout/Loadable';
import Loader from 'examples/ListLoader/Loadable';
import Register from 'containers/Auth/Registration/Loadable';
import ExportDataToCsv from 'examples/ExportDataToCsv/Loadable';
import Users from 'examples/Users/Loadable';
import Charts from 'examples/Charts/Loadable';
import Products from 'examples/Products/Loadable';
import SampleForm from 'examples/SampleForm/Loadable';
import ChangePassword from 'containers/ChangePassword/Loadable';
import MultiTabSupport from 'examples/MultiTabSupport/Loadable';
import ForgotPassword from 'containers/Auth/ForgotPassword/Loadable';
import NumeralConversion from 'examples/NumeralConversion/Loadable';
import ReactHookForm from 'examples/ReactHookForm/Loadable';
import { FAV_ICONS } from './constants';
import PrivateRoute from './PrivateRoute';
import RoleMiddleWare from './RoleMiddleWare';
import AuthRoute from './AuthRoute';
import GlobalStyle from '../../global-styles';
import { ROUTES } from '../constants';
import { manageSession } from '../../utils/Helper';
import { initGA, recordPageViewGA } from '../../utils/googleAnalytics';

const AppWrapper = styled.div`
  display: flex;
  min-height: 100vh;
  flex-direction: column;
`;

export default function App() {
  useEffect(() => {
    // google analytics init
    initGA();
    // first time page render
    recordPageViewGA(window.location.pathname);
    window.addEventListener('storage', manageSession, []);
    return () => {
      window.removeEventListener('storage', window);
    };
  }, []);

  useEffect(() => {
    // record page view on every route change
    recordPageViewGA(window.location.pathname);
  }, [window.location.pathname]);

  return (
    <AppWrapper data-testid="AppRoutes">
      <Helmet
        titleTemplate="%s - React.js Boilerplate"
        defaultTitle="React.js Boilerplate"
      >
        <meta name="description" content="A React.js Boilerplate application" />
        {FAV_ICONS.map((favIcon, index) => (
          <link {...favIcon} key={index} />
        ))}
      </Helmet>
      <Routes>
        {/* Private Routes */}
        <Route element={<PrivateRoute />}>
          <Route exact path={ROUTES.HOME} element={<FeaturePage />} />
          <Route path={ROUTES.GITHUB_SEARCH} element={<HomePage />} />
          <Route path={ROUTES.PROFILE} element={<Profile />} />
          <Route path={ROUTES.LOGOUT} element={<Logout />} />
          <Route path={ROUTES.LOADER} element={<Loader />} />
          <Route path={ROUTES.EXPORT_DATA} element={<ExportDataToCsv />} />
          <Route path={ROUTES.USERS} element={<Users />} />
          <Route path={ROUTES.CHARTS} element={<Charts />} />
          <Route path={ROUTES.PRODUCTS} element={<Products />} />
          <Route
            path={ROUTES.MULTI_TAB_SUPPORT}
            element={<MultiTabSupport />}
          />
          <Route path={ROUTES.CHANGE_PASSWORD} element={<ChangePassword />} />
          <Route path={ROUTES.SAMPLE_FORM} element={<SampleForm />} />
          <Route
            path={ROUTES.NUMERAL_CONVERTER}
            element={<NumeralConversion />}
          />
          <Route path={ROUTES.REACT_HOOK_FORM} element={<ReactHookForm />} />
        </Route>
        {/* RoleMiddleware */}
        <Route element={<RoleMiddleWare />}>
          <Route
            path={ROUTES.TEST_ADMIN_PAGE}
            element={() => <div>This is Admin Role Page</div>}
            // ShowError redirects to 403
            // showError
          />
        </Route>
        {/* Auth Routes */}
        <Route element={<AuthRoute />}>
          <Route exact path={ROUTES.LOGIN} element={<Login />} />
          <Route
            exact
            path={ROUTES.TWO_FACTOR_AUTHENTICATION}
            element={<TwoFactorAuthentication />}
          />
          <Route exact path={ROUTES.REGISTER} element={<Register />} />
          <Route
            exact
            path={ROUTES.FORGOT_PASSWORD}
            element={<ForgotPassword />}
          />
        </Route>
        <Route exact path={ROUTES.UNAUTHORIZED} component={UnauthorizedPage} />
        <Route path="" component={NotFoundPage} />
      </Routes>
      <GlobalStyle />
    </AppWrapper>
  );
}
