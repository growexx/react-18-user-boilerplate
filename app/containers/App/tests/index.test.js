import React from 'react';
import { Provider } from 'react-redux';
import { act, render } from '@testing-library/react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { MemoryRouter } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import { store } from 'configureStore';
import App from '../index';
import { ROUTES } from '../../constants';
import StorageService from '../../../utils/StorageService';
import { TOKEN_KEY } from '../../../utils/constants';

let globalStore;
const renderer = new ShallowRenderer();
const componentWrapper = () =>
  render(
    <Provider store={globalStore}>
      <IntlProvider locale="en">
        <MemoryRouter initialEntries={[ROUTES.TEST_ADMIN_PAGE]}>
          <App />
        </MemoryRouter>
      </IntlProvider>
    </Provider>,
  );

describe('<App />', () => {
  beforeAll(() => {
    globalStore = store;
    StorageService.set(TOKEN_KEY, 'TOKENVALUE');
  });
  it('should render and match the snapshot', () => {
    act(() => renderer.render(<App />));
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
  it('should render Div', () => {
    const { container } = componentWrapper();
    const element = container.firstElementChild;
    expect(element.tagName).toEqual('DIV');
  });
});
