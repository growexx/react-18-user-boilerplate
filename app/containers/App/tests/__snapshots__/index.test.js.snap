// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render and match the snapshot 1`] = `
<App__AppWrapper
  data-testid="AppRoutes"
>
  <HelmetWrapper
    defaultTitle="React.js Boilerplate"
    defer={true}
    encodeSpecialCharacters={true}
    titleTemplate="%s - React.js Boilerplate"
  >
    <meta
      content="A React.js Boilerplate application"
      name="description"
    />
    <link
      href="IMAGE_MOCK"
      rel="icon"
      sizes="512x512"
      type="image/png"
    />
    <link
      href="IMAGE_MOCK"
      rel="icon"
      sizes="192x192"
      type="image/png"
    />
    <link
      href="IMAGE_MOCK"
      rel="icon"
      sizes="32x32"
      type="image/png"
    />
    <link
      href="IMAGE_MOCK"
      rel="icon"
      sizes="16x16"
      type="image/png"
    />
    <link
      href="IMAGE_MOCK"
      rel="apple-touch-icon"
    />
  </HelmetWrapper>
  <Routes>
    <Route
      element={<PrivateRoute />}
    >
      <Route
        element={<Unknown />}
        exact={true}
        path="/"
      />
      <Route
        element={<Unknown />}
        path="/github-search"
      />
      <Route
        element={<Unknown />}
        path="/profile"
      />
      <Route
        element={<Unknown />}
        path="/logout"
      />
      <Route
        element={<Unknown />}
        path="/loader"
      />
      <Route
        element={<Unknown />}
        path="/export-data"
      />
      <Route
        element={<Unknown />}
        path="/users"
      />
      <Route
        element={<Unknown />}
        path="/charts"
      />
      <Route
        element={<Unknown />}
        path="/products"
      />
      <Route
        element={<Unknown />}
        path="/multi-tab-support"
      />
      <Route
        element={<Unknown />}
        path="/change-password"
      />
      <Route
        element={<Unknown />}
        path="/sample-form"
      />
      <Route
        element={<Unknown />}
        path="/number-converter-demo"
      />
      <Route
        element={<Unknown />}
        path="/react-hook-form"
      />
    </Route>
    <Route
      element={<Unknown />}
    >
      <Route
        element={[Function]}
        path="/admin"
      />
    </Route>
    <Route
      element={<AuthRoute />}
    >
      <Route
        element={<Unknown />}
        exact={true}
        path="/login"
      />
      <Route
        element={<Unknown />}
        exact={true}
        path="/two-factor-authentication"
      />
      <Route
        element={<Unknown />}
        exact={true}
        path="/register"
      />
      <Route
        element={<Unknown />}
        exact={true}
        path="/forgot-password"
      />
    </Route>
    <Route
      component={[Function]}
      exact={true}
      path="/403"
    />
    <Route
      component={[Function]}
      path=""
    />
  </Routes>
  <Memo(l) />
</App__AppWrapper>
`;
